### PR TITLE
298: Convert CreateWidgetEntityTest to MFTF

### PR DESCRIPTION
**Pull request by MilaLesechko**
### Description (*)
**Test Case**
[Magento\Widget\Test\TestCase\CreateWidgetsEntityTest](https://github.com/magento/magento-functional-tests-migration/tree/2.3-develop/dev/tests/functional/tests/app/Magento/Widget/Test/TestCase/CreateWidgetsEntityTest.php)
[Magento\Widget\Test\TestCase\CreateWidgetEntityTest](https://github.com/magento/magento-functional-tests-migration/tree/2.3-develop/dev/tests/functional/tests/app/Magento/Widget/Test/TestCase/CreateWidgetEntityTest.php)

**Variations to be converted:**

Variation | TestName
-- | --
CreateWidgetEntityTestVariation1 | AdminCreateWidgetCmsStaticBlockTest
CreateWidgetEntityTestVariation3 | AdminCreateWidgetResentlyViewedProductsTest
CreateWidgetEntityTestVariation4 | AdminCreateRecentlyComparedProductsTest
CreateWidgetEntityTestVariation6 | AdminCreateWidgetCatalogProductLinkTest
CreateWidgetEntityTestVariation7 | AssertWidgetWithDisabledBlockHtmlCacheTypeTest - bug
CreateWidgetEntityTestVariation8 | AssertWidgetWithInvalidatedBlockHtmlCacheTypeTest - bug
CreateWidgetEntityTestWithDifferentThemes | AdminCreateWidgetEntityTestWithDifferentThemesTest.xml

### Fixed Issues (if relevant)

1. magento/magento-functional-tests-migration#298: Convert CreateWidgetEntityTest to MFTF
2. magento/magento-functional-tests-migration#296: Convert CreateWidgetsEntityTest to MFTF

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Catalog/Test/Mftf/ActionGroup/AdminAnchorCategoryActionGroup.xml [hunk 1]
```diff
@@ -26,4 +26,8 @@
         <checkOption selector="{{CategoryDisplaySettingsSection.anchor}}" stepKey="enableAnchor"/>
         <click selector="{{AdminCategoryMainActionsSection.SaveButton}}" stepKey="saveSubCategory"/>
     </actionGroup>
+
```
Comment (lbajsarowicz) [inline app/code/Magento/Catalog/Test/Mftf/ActionGroup/AdminAnchorCategoryActionGroup.xml]: There can be only one Action Group per file

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            <stories value="Create Widget"/>
+            <title value="Create Widget Recently Compared Products"/>
+            <description value="Create Widget Recently Compared Products"/>
+            <testCaseId value="MAGETWO-27916"/>
+            <severity value="CRITICAL"/>
+            <group value="widget"/>
+            <group value="mtf_migrated"/>
+        </annotations>
+        <before>
+            <actionGroup ref="LoginAsAdmin" stepKey="loginAsAdmin"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]:
```suggestion
            <actionGroup ref="AdminLoginActionGroup" stepKey="loginAsAdmin"/>
```

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        <after>
+            <deleteData createDataKey="createCategory" stepKey="deleteCategory"/>
+            <deleteData createDataKey="simpleProduct1" stepKey="deleteProduct1"/>
+            <deleteData createDataKey="simpleProduct2" stepKey="deleteProduct2"/>
+            <deleteData createDataKey="simpleProduct3" stepKey="deleteProduct3"/>
+            <actionGroup ref="AdminDeleteWidgetActionGroup" stepKey="deleteWidget">
+                <argument name="widget" value="DefaultResentlyComparedProductsWidget"/>
+            </actionGroup>
+            <actionGroup ref="logout" stepKey="logoutOfAdmin"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]:
```suggestion
            <actionGroup ref="AdminLogoutActionGroup" stepKey="logoutOfAdmin"/>
```

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        </actionGroup>
+        <actionGroup ref="StorefrontAddProductToCompareActionGroup" stepKey="addToCompareProduct3">
+            <argument name="productVar" value="$$simpleProduct3$$"/>
+        </actionGroup>
+        <amOnPage url="{{StorefrontCategoryPage.url($$createCategory.name$$)}}" stepKey="onCategory"/>
+        <waitForPageLoad stepKey="waitForCategoryPageLoad"/>
+        <actionGroup ref="StorefrontClearCompareActionGroup" stepKey="clearCompare"/>
+
+        <amOnPage url="{{StorefrontHomePage.url}}" stepKey="amOnHomePage"/>
+        <waitForPageLoad stepKey="waitForPageLoad"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: Use existing Action Groups to load the storefront :-)

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        </actionGroup>
+
+        <actionGroup ref="OpenStoreFrontProductPageActionGroup" stepKey="openProduct3">
+            <argument name="productUrlKey" value="$$simpleProduct3.custom_attributes[url_key]$$"/>
+        </actionGroup>
+        <actionGroup ref="StorefrontAddProductToCompareActionGroup" stepKey="addToCompareProduct3">
+            <argument name="productVar" value="$$simpleProduct3$$"/>
+        </actionGroup>
+        <amOnPage url="{{StorefrontCategoryPage.url($$createCategory.name$$)}}" stepKey="onCategory"/>
+        <waitForPageLoad stepKey="waitForCategoryPageLoad"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: Never use `category.name` as a URL. There's `url_key` available as a special attribute.
The best approach is to use Action Group for opening category.

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        <fillField userInput="{{DefaultRecentlyViewedProductWidget.page_size}}" selector="{{WidgetSection.PageSize}}" stepKey="openSelectBlock"/>
+        <selectOption selector="{{AdminNewWidgetSection.productsAttributesToShow}}" parameterArray="[{{DefaultResentlyComparedProductsWidget.show_attributes[0]}},{{DefaultResentlyComparedProductsWidget.show_attributes[1]}},{{DefaultResentlyComparedProductsWidget.show_attributes[2]}}]" stepKey="selectAttributes"/>
+        <selectOption selector="{{AdminNewWidgetSection.buttonsToShow}}" parameterArray="[{{DefaultResentlyComparedProductsWidget.show_buttons[0]}}, {{DefaultResentlyComparedProductsWidget.show_buttons[1]}}, {{DefaultResentlyComparedProductsWidget.show_buttons[2]}}]" stepKey="selectButtons"/>
+        <click selector="{{AdminNewWidgetSection.saveWidget}}" stepKey="saveWidget"/>
+        <waitForPageLoad stepKey="waitForSaving"/>
+
+        <!-- Assert widget success save message -->
+        <see userInput="The widget instance has been saved." stepKey="assertWidgetSuccessSaveMessage"/>
+        <magentoCLI command="cache:flush" stepKey="flushCache"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: Could you replace that with `cache:clean` and specify the types of cache that are necessary to clean. For example that could be `cache:clean layout block_html full_page`

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            <argument name="widget" value="DefaultResentlyComparedProductsWidget"/>
+        </actionGroup>
+
+        <!-- Add widget options for resently viewed products widget -->
+        <waitForElement selector="{{WidgetSection.PageSize}}" stepKey="waitWidgetOptions"/>
+        <fillField userInput="{{DefaultRecentlyViewedProductWidget.page_size}}" selector="{{WidgetSection.PageSize}}" stepKey="openSelectBlock"/>
+        <selectOption selector="{{AdminNewWidgetSection.productsAttributesToShow}}" parameterArray="[{{DefaultResentlyComparedProductsWidget.show_attributes[0]}},{{DefaultResentlyComparedProductsWidget.show_attributes[1]}},{{DefaultResentlyComparedProductsWidget.show_attributes[2]}}]" stepKey="selectAttributes"/>
+        <selectOption selector="{{AdminNewWidgetSection.buttonsToShow}}" parameterArray="[{{DefaultResentlyComparedProductsWidget.show_buttons[0]}}, {{DefaultResentlyComparedProductsWidget.show_buttons[1]}}, {{DefaultResentlyComparedProductsWidget.show_buttons[2]}}]" stepKey="selectButtons"/>
+        <click selector="{{AdminNewWidgetSection.saveWidget}}" stepKey="saveWidget"/>
+        <waitForPageLoad stepKey="waitForSaving"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: Extract that to Action Group

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+        <!-- Assert widget success save message -->
+        <see userInput="The widget instance has been saved." stepKey="assertWidgetSuccessSaveMessage"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: Extract that to Action Group eg. `AssertAdminWidgetSavedSuccessfullyActionGroup`

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            <createData entity="_defaultCategory" stepKey="createCategory"/>
+            <createData entity="ApiSimpleProduct" stepKey="simpleProduct1">
+                <requiredEntity createDataKey="createCategory"/>
+            </createData>
+            <createData entity="ApiSimpleProduct" stepKey="simpleProduct2">
+                <requiredEntity createDataKey="createCategory"/>
+            </createData>
+            <createData entity="ApiSimpleProduct" stepKey="simpleProduct3">
+                <requiredEntity createDataKey="createCategory"/>
+            </createData>
```
Comment (lbajsarowicz) [inline app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: After creating products it's good practice to add `<magentoCron stepKey="runCronIndex" groups="index">` that for `On schedule`-based indices that triggers indices update.

File: app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml (hunk 7 above)
Comment (MilaLesechko) [reply app/code/Magento/Widget/Test/Mftf/Test/AdminCreateRecentlyComparedProductsTest.xml]: It is only one time used action. I don't think that this is needed

**Pull request merged**